---
title: "SJ Load"
author: OW
output: html_myeebook
editor_options: 
  chunk_output_type: console
---

```{r, warning=FALSE, message=FALSE, warning=FALSE, echo=F}
rm(list=ls())
ROOT_DIR<-"/fh/fast/furlan_s/user/owaltner/sj"
DATA_DIR <- file.path(ROOT_DIR, "data")      # SPECIFY HERE
RES_DIR  <- file.path(ROOT_DIR,  "res")     # SPECIFY HERE
RMD_DIR  <- file.path(ROOT_DIR,"rmd")     # SPECIFY HERE
CDS_DIR <- file.path(ROOT_DIR,   "cds")
FIG_DIR <- file.path(ROOT_DIR,  "figs")

suppressPackageStartupMessages({
  library(monocle3)
  library(Seurat)
  #library(m3addon)
  library(dplyr)
  library(Matrix)
  library(reticulate)
  library(ggplot2)
  library(pals)
  library(RColorBrewer)
  library(Seurat)
  library(ComplexHeatmap)
  library(ArchR)
  library(parallel)
  library(glmnet)
  library(ggplot2)
  library(BuenColors)
  library(cowplot)
  library(foreach)
  library(doMC)
  library(stringr)
  library(viridis)
  library(TCGAbiolinks)
  library(ComplexHeatmap)
  library(GenomicRanges)
  library(parallel)
  library(doMC)
  library(glmnet)
  library(pbapply)
  library(pbmcapply)
  library(ggplot2)
  library(cowplot)
  library(reshape2)
  library(gridExtra)
  library(edgeR)
  library(biomaRt)
  library(Matrix)
  library(scCustomize)
  library(clusterProfiler)
})
set.seed(1234)

library(future)
#version 11 cuda
dyn.load('/app/software/ArrayFire/3.8.1/lib64/libaf.so.3')

library(RcppArrayFire)
library(viewmaster)
```

```{r, load colors and object}
tum<-readRDS(file.path(ROOT_DIR, "cds/tumor_cellbender_filtered.rds"))#all cells post qc
cite<-readRDS("/fh/fast/furlan_s/user/owalt/ewings/cite/092921_cite.rds")
pdx<- readRDS( file.path(CDS_DIR, "pdx_samples.rds"))

tumor_cols<-c("#81c1a8", "#455d98", "#7253a3","#e07540","#e7a79c","#cf3732")
tum$tumor<-factor(tum$tumor, levels = c("tumor.1", "tumor.4", "tumor.2","tumor.3","tumor.5A", "tumor.5B"))
names(tumor_cols)<-levels(factor(tum$tumor))

rna_cols <- paletteContinuous(n=8)[c(1:3, 6:8)]

cite_cols= c("A4573" = "#9893DA","A673"="#ebeb78", "CHLA10" ="#fa9f42","CHLA9"="#810e44","PDX305"="#d94745","RDES"="#00B0E0","SKNMC"="#3e5e8e","TC32"="#0b6e4f","TC71" ="#b8e3ac")
pdx_cols<-paletteDiscrete(values = levels(factor(pdx$dataset)), set = "calm")
```

#make object
```{r}
files<-list.files(file.path(ROOT_DIR, "cellbender_filtered"))
paths<-paste0(ROOT_DIR,"/cellbender_filtered/" ,files[1:7])

# Load h5 files
sj_list <- lapply(paths, function(x){Read10X_h5(x)})

samps<-basename(files[1:7])

names(sj_list)<-samps

# Initialize the Seurat object with the cellbender normalized data.
seus<-lapply(1:length(samps), function(i){
  seu<-CreateSeuratObject(
    counts = sj_list[[i]])
  seu$dataset<-samps[i]
  seu
})

sj <- merge(
  x = seus[[1]],
  y = seus[2:length(seus)],
  add.cell.ids = samps
)
```

```{r}
DefaultAssay(sj) <- "RNA"
sj[["percent.mt"]] <- PercentageFeatureSet(sj, pattern = "^MT-")
sj$log_RNA<-log10(sj$nCount_RNA)

```

```{r RNA/ATAC QC}
VlnPlot(sj, features =  c("nCount_RNA" ,"percent.mt"), pt.size = 0, ncol = 3, log=T, group.by="dataset")
```

```{r RNA/ATAC QC}
#remove one run that's really poor quality
sj<-sj[,sj$dataset != samps[3]] #SJEWS049193_D2

sub<-sj[,sj$percent.mt < 15]
sub<-sub[,sub$log_RNA >3 &  sub$log_RNA <5.25]

VlnPlot(sub, features =  c("log_RNA" ,"percent.mt"), pt.size = 0, ncol = 3, log=T, group.by="dataset", raster =F)   

pdf("~/fig/plot.pdf", width = "9", height = )
VlnPlot(sub, features =  c("log_RNA" ,"percent.mt"), pt.size = 0, ncol = 3, log=T, group.by="dataset", raster =F)  
dev.off()
```

#normalize data
```{r}
sj <- NormalizeData(sub)
sj <- FindVariableFeatures(sj, selection.method = "vst", nfeatures = 2000)
sj <- ScaleData(sj, features = VariableFeatures(sj))

#reduce dims
sj <- SCTransform(sj, vars.to.regress = "percent.mt")
sj <- RunPCA(sj, features = VariableFeatures(sj))
ElbowPlot(sj, ndims = 30)
sj<-RunUMAP(sj, dims = 1:15, verbose = T, reduction.name = "SCT_UMAP") 
DimPlot(sj,  group.by ="dataset", label=T, reduction = "SCT_UMAP")+ggtitle("SCT UMAP")
library(harmony)
sj<- RunHarmony(
  object = sj,
  group.by.vars = 'dataset',
  reduction = 'pca',
  assay.use = 'SCT',
  project.dim = FALSE
)

sj <- RunUMAP(sj, dims = 1:15, reduction = 'harmony')
DimPlot(sj, group.by = "dataset", label=T, reduction = "umap")+ggtitle("Harmony integration")

FeaturePlot(sj, features= c("CD99",  "NR0B1", "NKX2-2", "PTPRC", "CD34", "ENG","SFTPB","ALPL", "COL1A1", "IBSP"), reduction = "umap", max.cutoff = "q99", order = T)
```

```{r}
aynaud<-read.delim("/fh/fast/furlan_s/user/owalt/ewings/multiome/EW4/ICEWS_genes.txt")
DefaultAssay(sj)<-"SCT"
sj<-AddModuleScore(sj, features =fus_kmean, name = "K")
sj<-AddModuleScore(sj, features = list(aynaud$Gene), name = "IC_EWS")

FeaturePlot_scCustom(sj,features = "IC_EWS1",reduction = "umap")&NoAxes()
```

#add metadata
```{r}
meta<-read.csv("~/sj/20220706_EWS_metadata.csv")

sj$id<-sapply(sj$dataset, function(x){
  t<-strsplit(x, "_")
  data<-paste0(t[[1]][1], "_",t[[1]][2])
  data
})

sj$sample<-sj$id
sj$sample[sj$id %in% c("SJEWS001321_D2", "SJEWS063829_D1")]<-"primary"
sj$sample[!sj$id %in% c("SJEWS001321_D2", "SJEWS063829_D1")]<-"metastasis"
table(sj$sample)

DimPlot(sj, group.by = "sample", reduction = "umap")

sj$sample_site<-sj$id
i<-levels(factor(sj$id))[1]
for(i in levels(factor(sj$id))){
  sj$sample_site[sj$id == i]<-meta[meta$SJID == i,]$Sample.Site
}
DimPlot(sj, group.by = "sample_site", reduction = "umap")

sj$primary_site<-sj$id
i<-levels(factor(sj$id))[1]
for(i in levels(factor(sj$id))){
  sj$primary_site[sj$id == i]<-meta[meta$SJID == i,]$Primary.Site
}

DimPlot(sj, group.by = "primary_site", reduction = "umap")

tum$tumor<-tum$id
tum$tumor[tum$id == "SJEWS001321_D2"]<- "tumor.1"
tum$tumor[tum$id == "SJEWS049193_D1"]<- "tumor.2"
tum$tumor[tum$id == "SJEWS063826_D1"]<- "tumor.3"
tum$tumor[tum$id == "SJEWS063829_D1"]<- "tumor.4"
tum$tumor[tum$id == "SJEWS063834_A1"]<- "tumor.5A"
tum$tumor[tum$id == "SJEWS063834_A2"]<- "tumor.5B"

```

```{r}
tum$subgroup_tum<-tum$tumor
tum$subgroup_tum[tum$tumor=="tumor.1" & tum$log_cell_type == "group.1"]<-"tum1_g1"

for(x in levels(factor(tum$tumor))){
  for(i in levels(factor(tum$log_cell_type))){
    tum$subgroup_tum[tum$tumor==x & tum$log_cell_type == i]<-paste0(x, "_" ,i)
  }
}
tum$subgroup_tum
```


#ID cell types
```{r}
sj<-FindNeighbors(sj, reduction = "umap", dims = 1:2)
sj<-FindClusters(sj, resolution = 0.025)
Idents(sj)<-"seurat_clusters"
markers<-FindAllMarkers(sj, only.pos = T)

top_n<- markers %>%
  group_by(cluster) %>%
  dplyr::filter(!str_detect(gene, "^RPL")) %>%
  dplyr::filter(!str_detect(gene, "^RPS")) %>%
  dplyr::filter(!str_detect(gene, "^MT")) %>%
  slice_max(n = 10, order_by = avg_log2FC) 
DoHeatmap(sj, features = top_n$gene)

q<-DimPlot(sj, reduction = "umap")
```

```{r}
DefaultAssay(sj)<-"RNA"
FeaturePlot_scCustom(sj, features = c("ALPL", "COL1A1", "IBSP"), reduction = "umap", order =T, colors_use = c("gray90",  "#04578f", "#801b1b"))#osteoblasts cluster 4
FeaturePlot_scCustom(sj, features = c("PLVAP", "COL4A1", "ENG"), reduction = "umap", order =T,  colors_use = c("gray90",  "#04578f", "#801b1b"))#blood vessel endothelium cluster 3
FeaturePlot_scCustom(sj, features = c("PTPRC"), reduction = "umap",order =T,  colors_use = c("gray90",  "#04578f", "#801b1b"))#immune cells 5
FeaturePlot_scCustom(sj, features = c("SFTPB"), reduction = "umap", order =T,  colors_use = c("gray90",  "#04578f", "#801b1b"))#lung tissue 6
FeaturePlot_scCustom(sj, features = c("CD99", "NKX2-2", "HOXD13", "NR0B1"), reduction = "umap", order =T,  colors_use = c("gray90",  "#04578f", "#801b1b"))#tumor markers 0,1,2
```

```{r}
p+q
```

```{r}
sj$cell_type<-as.character(sj$seurat_clusters)
sj$cell_type[sj$seurat_clusters %in% c(0,1,2)]<-"tumor"
# sj$cell_type[sj$seurat_clusters %in% c(5)]<-"smooth muscle"
sj$cell_type[sj$seurat_clusters %in% c(3)]<-"blood_vessel_endothelium"
sj$cell_type[sj$seurat_clusters %in% c(4)]<-"osteoblasts/chondrocytes"
sj$cell_type[sj$seurat_clusters %in% c(5)]<-"immune_cells"
sj$cell_type[sj$seurat_clusters %in% c(6)]<-"lung_tissue"

DimPlot(sj, group.by = "cell_type", reduction = "umap")
```

#make monocle object
```{r}
cds<-seurat_to_monocle3(sj, seu_rd= "umap")
plot_cells(cds, color_cells_by = "cell_type", label_cell_groups = F, cell_size = 0.75)+scale_color_manual(values=ArchR::paletteDiscrete(values = levels(factor(sj$cell_type)), set = "grove"))+NoAxes()
plot_cells(cds, color_cells_by = "Sample", label_cell_groups = F, cell_size = 0.5)+scale_color_manual(values=sfc(n=20, scramble = F))+NoAxes()
sj$log_RNA<-log10(sj$nCount_RNA)
VlnPlot(sj, features = "log_RNA", group.by = "Sample", cols =sfc(n=20, scramble = F), pt.size = 0)

plot_cells(cds, genes = "MKI67", label_cell_groups = F, cell_size = 0.75, scale_to_range = F)+scale_color_gradientn(colors=c("gray90",  "#04578f", "#801b1b"))+NoAxes()
```

#tumor only patient samples UMAP, Fig 5A
```{r}
tum<-sj[,sj$cell_type == "tumor"]
set.seed(1234)
tum <- NormalizeData(tum, normalization.method = "LogNormalize", scale.factor = 10000)
tum <- FindVariableFeatures(tum, selection.method = "vst", nfeatures = 2000)
tum <- ScaleData(tum, features = VariableFeatures(tum))

#reduce dims
tum <- SCTransform(tum, vars.to.regress = c("log_RNA"))
tum <- RunPCA(tum, features = VariableFeatures(tum))
ElbowPlot(tum, ndims = 30)
tum<-RunUMAP(tum, dims = 1:25, verbose = T, reduction.name = "SCT_UMAP") 
tum<-FindNeighbors(tum)
tum<-FindClusters(tum, resolution = 0.6)

DimPlot(tum)

VlnPlot(tum, features = "log_RNA", group.by = "tumor", pt.size =0, cols = c( brewer.spectral(n=10)[c(8,3,5,10)],brewer.paired(n=9)[5:6]))
DimPlot(tum, group.by = "dataset", label=T, reduction = "SCT_UMAP")

cds<-seurat_to_monocle3(tum, seu_rd= "SCT_UMAP")

plot_cells(cds, color_cells_by = "dataset", label_cell_groups = F, cell_size = 0.75)+scale_color_manual(values=c( brewer.spectral(n=10)[c(8,3,5,10)],brewer.paired(n=9)[5:6]))+NoAxes()
```



#Fig 5B & S5B
```{r fusion dependent kmean genes UMAP}
fus_kmean<-readRDS("/fh/fast/furlan_s/user/owaltner/m3/res/kd_kmean_genes.rds")
aynaud<-read.delim("/fh/fast/furlan_s/user/owalt/ewings/multiome/EW4/ICEWS_genes.txt")
DefaultAssay(tum)<-"SCT"
tum<-AddModuleScore(tum, features =fus_kmean, name = "K")
tum<-AddModuleScore(tum, features = list(aynaud$Gene), name = "IC_EWS")

pdf("~/sj/figs/kmean_umap.pdf", width = 7, height = 7)
FeaturePlot_scCustom(tum,features = c("K1", "K2", "K3"), colors_use = rna_cols , reduction = "SCT_UMAP")&NoAxes()
dev.off()

pdf("~/sj/figs/icews_umap.pdf", width = 7, height = 7)
FeaturePlot_scCustom(tum,features = c("IC_EWS1"), colors_use = rna_cols , reduction = "SCT_UMAP")&NoAxes()
dev.off()
```



#Save object
```{r}
saveRDS(sj, file.path("~/sj/cds", "sj_cellbender_filtered.rds"))
saveRDS(tum, file.path("~/sj/cds", "tumor_cellbender_filtered.rds"))
```
## Make count matrix
```{r load objects}
dds<-readRDS(file.path(CDS_DIR, "bulkRNA_dds.rds"))
```

```{r Get gene lengths}
counts<-dds@assays@data$normalized_counts
rownames(counts)<-rownames(counts) %>% strsplit("\\.") %>% sapply("[[", 1)
colnames(counts)<-colData(dds)$sample

library(data.table)
gtf<-fread("/Users/owaltner/Fred Hutchinson Cancer Research Center/Furlan_Lab - General/resources/gencode_v33.gtf")
gtf <- rtracklayer::import('/Users/owaltner/Fred Hutchinson Cancer Research Center/Furlan_Lab - General/resources/gencode_v33.gtf')
gtf_df <- as.data.frame(gtf)

mcols(dds)<-cbind(mcols(dds), gtf_df[match(mcols(dds)$gene_id, gtf_df$gene_id),])

mcols(dds)$size<-mcols(dds)$end - mcols(dds)$start
towrite<-cbind(dds@assays@data$normalized_counts)
towrite<-cbind(data.frame(size=mcols(dds)$size), towrite)
cn<-colnames(towrite)
rownames(towrite)<-make.unique(mcols(dds)$gene_id %>% strsplit("\\.")  %>% sapply("[[", 1))
head(towrite)
colData(dds)

write.csv(towrite, file.path(DATA_DIR, "bulk_counts_normalized.csv"), row.names = rownames(towrite))
write.csv(colData(dds), file.path(DATA_DIR, "bulk_meta.csv"))
```

#read/ prepare objects
```{r prepare cite object}
mo<-readRDS("~/m2/m2.rds")
cite<-readRDS("/fh/fast/furlan_s/user/owalt/ewings/cite/210330_cds_filtering.RDS")
```

```{r prepare cite object}
cite$cell_line<-clusters(cite)
cite$cell_line<-as.character(cite$cell_line)
cite$cell_line[which(clusters(cite) == 1)]<-"A673"
cite$cell_line[which(clusters(cite) == 2)]<-"PDX305"
cite$cell_line[which(clusters(cite) == 3)]<-"CHLA10"
cite$cell_line[which(clusters(cite) == 4)]<-"TC71"
cite$cell_line[which(clusters(cite) == 5)]<-"TC32"
cite$cell_line[which(clusters(cite) == 6)]<-"SKNMC"
cite$cell_line[which(clusters(cite) == 7)]<-"CHLA9"
cite$cell_line[which(clusters(cite) == 8)]<-"A4573"
cite$cell_line[which(clusters(cite) == 9)]<-"RDES"


cite$subgroup<-clusters(cite)
cite$subgroup<-as.character(cite$subgroup)
cite$subgroup[which(clusters(cite) == 1)]<-"k1"
cite$subgroup[which(clusters(cite) == 2)]<-"k2"
cite$subgroup[which(clusters(cite) == 3)]<-"k2"
cite$subgroup[which(clusters(cite) == 4)]<-"k3"
cite$subgroup[which(clusters(cite) == 5)]<-"k2"
cite$subgroup[which(clusters(cite) == 6)]<-"k3"
cite$subgroup[which(clusters(cite) == 7)]<-"k2"
cite$subgroup[which(clusters(cite) == 8)]<-"k3"
cite$subgroup[which(clusters(cite) == 9)]<-"k3"

exprs <- exprs(cite)
meta <- colData(cite) %>% as.data.frame()
embedding <- cite@int_colData@listData[["reducedDims"]]@listData[["UMAP"]] %>% as.data.frame()
feature_meta <-  rowData(cite) %>% as.data.frame()
names(feature_meta)
```

```{r set up python enrionment}
Sys.setenv(RETICULATE_PYTHON = "~/.conda/envs/py3/bin/python3")
library("SeuratDisk")
library("Seurat")
library("reticulate")
use_python("~/.conda/envs/py3/bin/python3")
use_condaenv("~/.conda/envs/py3")
py_config()
```

#python libraries
```{python }
#############
# Libraries #
#############
import os
import datetime
import argparse
#Suppress lots of mecitery warning
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' 
from os.path import expanduser as eu
from os import path
from math import ceil
import tensorflow as tf
import scanpy as sc
import pandas as pd
import numpy as np
from scipy.stats import poisson 
from scipy.special import loggamma
from scipy.io import mmread
import scipy.sparse as sparse
#from pandas.core.algorithms import match
import warnings
from matplotlib.pyplot import rc_context
tf.compat.v1.disable_eager_execution() 
```

#make scanpy objects
```{python 9 CL h5ad}
adata= sc.AnnData(X = r.exprs.T, obs = r.meta, var = r.feature_meta)
adata.obsm['X_umap'] = r.embedding
adata.raw = adata # raw counts
# adata.write('/fh/fast/furlan_s/user/owalt/ewings/bulk_RNA_patient/data/snRNA_EwS.h5ad')
# adata.write('/fh/fast/furlan_s/user/owalt/ewings/bulk_RNA_patient/data/scRNA_EwS.h5ad')
```

#run cell signal analysis
```{python}

################
# modify These #
################

os.chdir("/home/owaltner/dev/cellSignalAnalysis")
single_cell_ref = adata
single_cell_ref_column = "subgroup"
bulk_count_data = "/fh/fast/furlan_s/user/owalt/ewings/bulk_RNA_patient/data/bulk_counts_normalized.csv"
bulk_meta_data = "/fh/fast/furlan_s/user/owalt/ewings/bulk_RNA_patient/data/bulk_meta.csv"
gene_weights = "/home/owaltner/dev/cellSignalAnalysis/geneWeights.tsv"

ds = range(0,126)
drop_zero_sig_genes=False
init_log_exposure=-10
insert_size=500
l1_lambda=0
l2_lambda=0
learn_rate=0.01
log_likelihood_tolerance=1e-06
max_it=10000000.0
no_collapse_ref=False
poll_interval=100
refine_sigs=None
sparsity_tolerance=0.0001
```

```{python}
####################
# Load scReference #
####################
sc.pl.umap(adata, color = single_cell_ref_column)
```

```{python step 1}
##################
# Load bulk data #
##################
toc = pd.read_csv(bulk_count_data, index_col=0)
tol=toc['size']
toc = toc.drop(columns='size')
meta_bulk = pd.read_csv(bulk_meta_data, index_col=0)
bulkGenes = toc.index
toc = toc.iloc[:,ds]
meta_bulk = meta_bulk.iloc[ds,:]
n=toc.shape[1]
tol = pd.concat([tol] * n, axis=1, ignore_index=True)
toc.index.name="geneName"
tol.index.name="geneName"
tol.columns=toc.columns
dat = adata.raw.X.T
cellSigs = np.asarray(adata.obs[single_cell_ref_column])
tmp = []
scSigs=[]
for cellSig in list(set(adata.obs[single_cell_ref_column])):
    idxs = np.where(cellSigs==cellSig)[0]
    tgts = dat[:,idxs]
    #Collapse across cells
    tgts = tgts.sum(axis=1)
    #Normalise to sum to 1 and store
    tmp.append(tgts/tgts.sum())
tmp = pd.DataFrame([np.squeeze(np.asarray(x)) for x in tmp],index=list(set(cellSigs)),columns=adata.var['id']).transpose()
scSigs.append(tmp)
refGenes = list(set.intersection(*[set(x.index.values) for x in scSigs]))
refGenes.sort()
nGenes = [x.shape[0] for x in scSigs]
if len(refGenes) < max(nGenes):
  warnings.warn("Reference cellular signals have between %d and %d genes, with only %d in comciten.  All non-shared genes will be dropped."%(min(nGenes),max(nGenes),len(refGenes)))
#Merge into one matrix 
scSigs = [x.loc[refGenes] for x in scSigs]
scSigs = pd.concat(scSigs,axis=1)

################
# Load weights #
################
if gene_weights is not None:
  geneWeights = pd.read_csv(eu(gene_weights),sep='\t',index_col='geneID')
else:
  geneWeights = pd.DataFrame(np.ones((toc.shape[0],1)),index=toc.index,columns=['weight'])


##################
# Harmonise data #
##################

commonGenes = list(set.intersection(set(refGenes), set(geneWeights.index), set(bulkGenes)))
if len(commonGenes) < max(len(bulkGenes),len(refGenes)):
  warnings.warn("Bulk data has %d genes, reference data %d, with only %d in common.  All information relating to non-shared genes will be dropped."%(len(bulkGenes),len(refGenes),len(commonGenes)))
#Do the more stringent thing of dropping anything zero in sigs
if drop_zero_sig_genes:
  print("Dropping %d genes with zero expression in any of the supplied cellular signals. %d genes remain"%(sum(scSigs.loc[commonGenes].sum(axis=1)==0),sum(scSigs.loc[commonGenes].sum(axis=1)>0)))
  commonGenes = list(np.array(commonGenes)[(scSigs.loc[commonGenes].sum(axis=1))>0])
geneWeights = geneWeights.loc[commonGenes]
tol = tol.loc[commonGenes]
scSigs = scSigs.loc[commonGenes]
#Finalise weights
geneWeights = geneWeights.reindex(commonGenes)
geneWeights = geneWeights.fillna(1.0)
scSigs = scSigs.loc[commonGenes]
toc = toc.loc[commonGenes,]
toc = toc.iloc[[list(toc.index.values).index(i) for i in commonGenes]]
#Re-normalise signals?
#scSigs = scSigs/scSigs.sum(axis=0)
#Some useful messages
print("Fitting %d bulk samples to %d cellular signals using %d genes"%(toc.shape[1],scSigs.shape[1],toc.shape[0]))
print("Signals named:")
tmp = scSigs.columns.values
print(str(tmp))
print("Regularising with lambda %g/%g for L1/L2 "%(l1_lambda, l2_lambda))
```

```{python step 2}
####################
# Define the model #
####################
p,n = toc.shape
s = scSigs.shape[1]
#########
# Inputs
S = tf.constant(scSigs.values.astype('float32'),name='fixed_signals')
#Note the insert-size conversion factor to convert length of gene to expected number of fragments per molecule of mRNA from gene
C = tf.constant((tol.values/insert_size).astype('float32'),name='mols_to_reads')
k = tf.constant(toc.values.astype('float32'),name='obs_counts')
w = tf.constant(geneWeights.weight.values.astype('float32'),name='gene_weights')
#Create "Signal" that is really just the intercept term
Si = tf.constant((np.ones((p,1))/p).astype('float32'),name='intercept_signal')
Sp = tf.concat([S,Si],axis=1)
############
# Exposures
#These are the things we actually train
#Initialise to -10000, which converts to 0.  This value will not be changed when fitting null
z = tf.Variable(tf.zeros([s,n])+init_log_exposure,name='exposures')
#Define dynamic intercept
int0 = tf.Variable(tf.zeros([1,n]),name='intercept')
#Merge
zz = tf.concat([z,int0],name='exposures_with_int',axis=0)
#Positive exposures, including intercept
E = tf.exp(zz,name='positive_exposures')
###################
# Predicted Counts
#Predicted number of molecules
q = tf.matmul(Sp,E,name='pred_mols')
#Convert to number of reads
y = tf.multiply(q,C,name='pred_reads')
##########################
# Poisson log-likelihood
#Variable part of Poisson log-likelihood
Dij = k*tf.math.log(y)- y
#Constant part
D0 = tf.math.lgamma(k+1)
#Add gene weights and sum to get negative log-likelihood
LL = tf.transpose(a=D0-Dij)*w
tLL = tf.reduce_sum(input_tensor=LL,name='NLL')
#############################
# Define objective function
#Count of non-zeroish co-efficients
cCnt = tf.reduce_sum(input_tensor=tf.sigmoid(zz))
# The final penalised, adjust NLL
O = tLL
############
# Optimiser
opt_E = tf.compat.v1.train.AdamOptimizer(learn_rate)


############
# Fit NULL #
############
print('''
######################
# Fitting null Model #
######################
''')
#Fit null first so we can inform the regularisation
update_op = z.assign(np.zeros([s,n])-1000)
learners = opt_E.minimize(O,var_list=[int0],name='learn_exposures')
#Initialise
nullSess = tf.compat.v1.Session()
init = tf.compat.v1.global_variables_initializer()
nullSess.run(init)
_ = nullSess.run(update_op)
#Record initial values
last = nullSess.run(O)
lastNonZero = nullSess.run(cCnt)
#Record the movements
null_nll = np.zeros(int(ceil(max_it/poll_interval)))
null_nsigs = np.zeros(int(ceil(max_it/poll_interval)))
i=0
while True:
  #Take the exposure step
  nullSess.run([learners])
  #Every now and then, record our progress and check if we've converged
  if i%poll_interval == 0:
    ii = i//poll_interval
    #Record object function and number of non-zero exposures
    null_nll[ii] = nullSess.run(O)
    null_nsigs[ii] = nullSess.run(cCnt)
    #Record how much we've changed since we last checked
    diff = (last-null_nll[ii])
    last = null_nll[ii]
    diffCnts = (lastNonZero - null_nsigs[ii])
    lastNonZero = null_nsigs[ii]
    #Calculate per-element summaries
    sigsPerSample = lastNonZero/n
    llPerEntry = null_nll[ii]/n/p
    #And the average intercept
    avgInt = np.mean(np.exp(nullSess.run(int0)))
    #The average coverage relative to the observed
    avgCov = np.mean(nullSess.run(y).sum(axis=0)/nullSess.run(k).sum(axis=0))
    print("[%s] step %d, training O=%g, cnt=%g,dO=%g,dCnt=%g,nSigsAvg=%g/%d,avgNLL=%g,avgCov=%g" %(datetime.datetime.now(),i,last,lastNonZero,diff,diffCnts,sigsPerSample,s+1,llPerEntry,avgCov))
    #Test if we should terminate
    if diff<=0 and diff/last < log_likelihood_tolerance  and diffCnts/lastNonZero < sparsity_tolerance:
      break
  i = i+1
  if i>max_it:
    break
```

```{python step 3}
################
# Update model #
################
#Get the NULL likelihoods per samples
sampL = nullSess.run(LL).sum(axis=1)
#And the penalty under the assumption of uniform exposures, that total to the number of reads
#This is the average number of molecules per sample
mols_per_samp = (toc/tol).sum(axis=0)*insert_size
#NOTE: This sets the scale for lambda so that for lambda=1 the sample -LL under the NULL model is equal to the imposed penalty if exposures are uniformly distributed.
#The L1 Penalty
#Construct the total penalty under the uniformity and avg mol assumptions.  The denominator is really s*(mols_per_samp/s)
avg_pen = mols_per_samp
avg_pen = sampL/avg_pen
#And the penalty itself
penL1 = l1_lambda*tf.reduce_sum(avg_pen.values*E)
#The L2 penalty, does not include lambda (we don't want to penalise large exposures in intercept)
#Under the uniformity assumption and sum to avg mol assumption, this is the total penalty
#NOTE: The s+1 instead of s is to account for there being the unpenalized intercept in the model
avg_pen = s*((mols_per_samp/(s+1))**2)
avg_pen = sampL/avg_pen
#Don't want to include the intercept so have to re-exponetiate
penL2 = tf.exp(z)**2
penL2 = l2_lambda*tf.reduce_sum(avg_pen.values*penL2)
#And the penalised likelihood
O_pen = O+penL1+penL2


############
# Main fit #
############
print('''
######################
# Fitting main Model #
######################
''')
#Define optimisers
toLearn = [z,int0]
learners = opt_E.minimize(O_pen,var_list=toLearn,name='learn_exposures')
#Initialise
sess = tf.compat.v1.Session()
init = tf.compat.v1.global_variables_initializer()
sess.run(init)
#Record initial values
last = sess.run(O_pen)
lastNonZero = sess.run(cCnt)
#Record the movements
nll = np.zeros(int(ceil(max_it/poll_interval)))
nsigs = np.zeros(int(ceil(max_it/poll_interval)))
i=0
while True:
  #Take the exposure step
  sess.run([learners])
  #Every now and then, record our progress and check if we've converged
  if i%poll_interval == 0:
    ii = i//poll_interval
    #Record object function and number of non-zero exposures
    nll[ii] = sess.run(O_pen)
    nsigs[ii] = sess.run(cCnt)
    #Record how much we've changed since we last checked
    diff = (last-nll[ii])
    last = nll[ii]
    diffCnts = (lastNonZero - nsigs[ii])
    lastNonZero = nsigs[ii]
    #Calculate per-element summaries
    sigsPerSample = lastNonZero/n
    llPerEntry = nll[ii]/n/p
    #And the average intercept
    avgInt = np.mean(np.exp(sess.run(int0)))
    #The average coverage relative to the observed
    avgCov = np.mean(sess.run(y).sum(axis=0)/sess.run(k).sum(axis=0))
    print("[%s] step %d, training O=%g, cnt=%g,dO=%g,dCnt=%g,nSigsAvg=%g/%d,avgNLL=%g,avgCov=%g" %(datetime.datetime.now(),i,last,lastNonZero,diff,diffCnts,sigsPerSample,s+1,llPerEntry,avgCov))
    #Test if we should terminate
    if diff<0 and diff/last < log_likelihood_tolerance  and diffCnts/lastNonZero < sparsity_tolerance:
      break
  i = i+1
  if i>max_it:
    break

```

```{python step 4}
###################
# Post processing #
###################
#First save the things used to fit.
#The bulk counts
#toc.to_csv(args.output + '_usedBulkCounts.tsv',sep='\t',index_label=False)
#And their lengths
#tol.to_csv(args.output + '_usedBulkGeneLengths.tsv',sep='\t',index_label=False)
#The cellular Signals
#scSigs.to_csv(args.output + '_usedCellularSignals.tsv',sep='\t',index_label=False)
#And gene weights
#geneWeights.to_csv(args.output + '_usedGeneWeights.tsv',sep='\t',index_label=False)
#Now the things we infered
#The exposures, plus goodness of fit metrics
pred_E = pd.DataFrame(sess.run(E),index=list(scSigs.columns)+['Intercept'],columns=toc.columns)
pR2 = 1-sess.run(LL).sum(axis=1)/nullSess.run(LL).sum(axis=1)
pred_E.loc['pR2'] = pR2
pred_E.loc['fitCount'] = sess.run(y).sum(axis=0)
pred_E.loc['obsCount'] = sess.run(k).sum(axis=0)
#pred_E.to_csv(args.output + '_fitExposures.tsv',sep='\t',index_label=False)
#Full predicted table of counts
pred_y = pd.DataFrame(sess.run(y),index=toc.index,columns=toc.columns)
#pred_y.to_csv(args.output + '_fitCounts.tsv',sep='\t',index_label=False)
#Full negative log-likelihood table
pred_LL = pd.DataFrame(sess.run(LL),index=toc.columns,columns=toc.index)
#pred_LL.to_csv(args.output + '_negativeLogLikelihoodFullFit.tsv',sep='\t',index_label=False)
#Full negative log-likelihood table under null
pred_nLL = pd.DataFrame(nullSess.run(LL),index=toc.columns,columns=toc.index)
#pred_nLL.to_csv(args.output + '_negativeLogLikelihoodNullFit.tsv',sep='\t',index_label=False)
#Stats about fitting itself
#Main fit
df = pd.DataFrame(np.c_[nll,nsigs],columns=['NLL','numExposures'])
df = df.iloc[np.logical_and(nll!=0,nsigs!=0),]
df.loc[:,'step'] = df.index*poll_interval
df.loc[:,'fitType']='FullFit'
fitStats = df
#Null Fit
df = pd.DataFrame(np.c_[null_nll,null_nsigs],columns=['NLL','numExposures'])
df = df.iloc[np.logical_and(null_nll!=0,null_nsigs!=0),]
df.loc[:,'step'] = df.index*poll_interval
df.loc[:,'fitType']='NullFit'
fitStats = pd.concat([fitStats,df])

```
#Visualize Data
```{python visualize}
###################
#    Plotting     #
###################
def normaliseExposures(fit):
    #Is this just the base?
    #fit = pd.read_csv(tgt,sep='\t')
    fit = fit[sorted(fit)]
    #Extract the goodness of fit rows
    gofNoms = ['pR2','fitCount','obsCount']
    gof = fit.loc[gofNoms]
    gof.loc['log2(countRatio)'] = np.log2(gof.loc['fitCount']/gof.loc['obsCount'])
    exposures = fit.loc[~fit.index.isin(gofNoms)]
    exposures = exposures/exposures.sum(0)
    #Done
    return {'exposures': exposures, 'gof': gof,'raw':fit.loc[~fit.index.isin(gofNoms)]}

#target_meta = pd.read_csv('./bulkData/SangerProjectTARGET_ALL_phase1.txt', index_col=0, header=None)
fit = normaliseExposures(pred_E)
fit = fit['exposures']
fit.columns = pd.Series(fit.columns).str.split('.', expand = True, n=1)[0].values
adata_target = sc.AnnData(fit.T)
#adata_target = adata_target[adata_target.obs_names.isin(target_meta.index.values)]
vn = pd.Series(adata_target.var_names)
adata_target.obs['sample']=adata_target.obs_names

```

#Fig S5E
```{r heatmaps cell line v cell line}
fit<-py$fit
fitW<-fit[rowSums(fit)>0.3,]

# fit_cl<-read.table(file.path(RES_DIR, "cl_to_cl_fit.mtx"), header = T, row.names = T)

colData(dds)$type<-colData(dds)$Group
colData(dds)$type[colData(dds)$Group != "Primary"]<-"cell_line"

cl_dds<-dds[,colData(dds)$type == "cell_line"]

cl_dds<-cl_dds[,colData(cl_dds)$Group %in% mo$cell_line]
table(cl_dds$sample, cl_dds$Group)
fit_cl<-fitW[,colnames(cl_dds)]
colnames(fit_cl)<-make.unique(cl_dds$Group)
cl_meta<-colData(cl_dds)

#write.table(fit_cl , file.path(RES_DIR, "cl_to_cl_fit.mtx"), col.names = T, row.names= T)

annotation_col=data.frame(subtype=colnames(fit_cl), row.names=colnames(fit_cl))
table(cl_meta$Group)
cols = list(subtype = c("A673"="#9893DA", "A673.1"="#9893DA","CHLA9" = "#ebeb78", "RDES" = "#b8e3ac", "RDES.1" = "#b8e3ac","TC32" = "#810e44", "TC32.1" = "#810e44","TC71" = "#0b6e4f", "SKNMC" ="#00B0E0" ))

pdf(file.path(FIG_DIR, "cl_to_cl_deconv.pdf"), width = 7, height = 5)
pheatmap(fit_cl[,c(2,4,7,1,5,8,9,3,6)], color = viridis(100), border_color =NULL, annotation_col = annotation_col, annotation_colors = cols, cluster_rows = F, cluster_cols = F, name = "Similarity Score", main= "Bulk RNAseq EWS Cell Line Deconvelution with scRNAseq")
dev.off()
```

```{r heatmaps cell line v cell line}
Heatmap(fit_cl)

cl<-rownames(fit_cl)
names(cl)<-rownames(fit_cl)
ha<-rowAnnotation(cell_line = cl[c("SKNMC", "A673", "CHLA9", "PDX305", "TC32", "A4573", "CHLA10", "TC71", "RDES", "Intercept")], col = list(cell_line = c("A4573"="#D51F26","A673"="#272E6A","CHLA9"="#208A42","CHLA10"="#89288F","PDX305"="#F47D2B", "RDES"="#FEE500","SKNMC"="#8A9FD1","TC32"="#3BBCA8","TC71"="#E6C2DC", "Intercept" = "gray" )))

cl<-colnames(fit_cl)
names(cl)<-colnames(fit_cl)
ca<-columnAnnotation(cell_line = cl[c("SKNMC", "A673", "A673.1", "CHLA9", "TC32","TC32.1", "TC71", "RDES", "RDES.1")], col = list(cell_line = c("A673"="#272E6A", "A673.1"="#272E6A","CHLA9" = "#208A42", "RDES" = "#FEE500", "RDES.1" = "#FEE500","TC32" = "#3BBCA8", "TC32.1" = "#3BBCA8","TC71" = "#E6C2DC", "SKNMC" ="#8A9FD1" ) ))

Heatmap(fit_cl[c("SKNMC", "A673", "CHLA9", "PDX305", "TC32", "A4573", "CHLA10", "TC71", "RDES", "Intercept"),c("SKNMC", "A673", "A673.1", "CHLA9", "TC32","TC32.1", "TC71", "RDES", "RDES.1")], cluster_rows = F,  cluster_columns = F, col = viridis(100), left_annotation = ha, top_annotation = ca)

```

#Fig 5C
```{r tumor vs subgroup}
sfit<-py$fit
# sfitW<-sfit[rowSums(sfit)>0.3,]

colData(dds)$type<-colData(dds)$Group
colData(dds)$type[colData(dds)$Group != "Primary"]<-"cell_line"

sub_dds<-dds[,colData(dds)$Group == "Primary"]

fit_subgroup<-sfit[,colnames(sub_dds)]
p_meta<-colData(sub_dds)

#write.matrix(fit_subgroup , file.path(RES_DIR, "subgroup_primary_tumor_fit.mtx"))

rownames(fit_subgroup)<-c("k2", "k1", "k3", "intercept")

cols = list(BioProject = c(PRJNA253759="gray90", PRJNA261990="#000000"))


bioproject<-unlist(p_meta$BioProject)
names(bioproject)<-colnames(fit_subgroup)

ca<-columnAnnotation(BioProject = bioproject, col = list(BioProject = c(PRJNA253759="#B0C4DE", PRJNA261990="#4682B4")), border = F)

Heatmap(fit_subgroup[c("k1", "k2", "k3", "intercept"),], col = viridis(100),  show_column_names  = F, show_row_names = T, cluster_rows = F, top_annotation = ca)
```

#Patient Microarray data set1
```{r helpful functions}
# gene_list<-list(kmean_probes[[3]])
# name="k2"
strat_curve<-function(obj, gene_list, name){
  obj<-AddModuleScore(obj, features = gene_list, name = name)
  quants<-quantile(obj[[paste0(name,"1")]] %>% unlist(), 0.85)
  obj$pos<-T
  obj$pos[obj[[paste0(name,"1")]] < quants]<-F
  
  cox<-coxph(Surv(EFS.time..days.,EFS..censor.0..event.1.) ~pos, data = obj@meta.data %>% data.frame()) %>% summary()
  pval<-cox$waldtest["pvalue"]

  survfit2(Surv(EFS.time..days.,EFS..censor.0..event.1.) ~ pos, data = obj@meta.data %>% data.frame()) %>% 
  ggsurvfit() +
  labs(
    x = "Days",
    y = "Event free survival probability"
  )+ylim(c(0,1))+ggtitle(label = paste0(name, " high expression / pvalue = " , round(pval, 4)))
}

gene<-"PTMA"
strat_curve_gene<-function(obj, gene){
  probes<-fdat[which(fdat$gene_short_name %in% gene), ]$id
  if(length(probes) >1){
    obj<-AddModuleScore(obj, features = list(probes), name = gene)
    quant<-quantile(obj[[paste0(name,"1")]] %>% unlist(), 0.5)
    obj$pos<-T
    obj$pos[obj[[paste0(name,"1")]] < quant]<-F  
  }
  if(length(probes) == 1){
    obj$gene<-obj@assays$RNA@counts[probes,]
    quant<-quantile(obj$gene %>% unlist(), 0.5)
    obj$pos<-T
    obj$pos[obj$gene < quant]<-F  
  }
  cox<-coxph(Surv(EFS.time..days.,EFS..censor.0..event.1.) ~pos, data = obj@meta.data %>% data.frame()) %>% summary()
  pval<-cox$waldtest["pvalue"]
  survfit2(Surv(EFS.time..days.,EFS..censor.0..event.1.) ~ pos, data = obj@meta.data %>% data.frame()) %>% 
  ggsurvfit() +
  labs(
    x = "Days",
    y = "Event free survival probability"
  )+ylim(c(0,1))+ggtitle(label = paste0(gene, " high expression / pvalue = " , round(pval, 4)))
}
```

```{r set up file paths}
# -------------------------------------

# -------------------------------------
# find relative directory on Server
# -------------------------------------

key<-"/fh/fast/furlan_s/user/owalt/ewings"
flag<-"EwingMA"
setwd(file.path(key, flag))
# -------------------------------------
# specify paths and load functions
# -------------------------------------
DATA_DIR <- file.path( key,flag,  "CEL") # SPECIFY HERE
RES_DIR  <- file.path(  key,flag, "res")      # SPECIFY HERE

directory.wb<-read.xlsx(file.path(key,flag, "excel_sheets/46 tumors with array profiles_BPC and REG numbers.xlsx"))

directory.ws<-directory.wb
colnames(directory.ws)<-gsub("\\.", "", colnames(directory.ws))
classvec=as.factor(directory.ws$`EWS-FLItranslocationtype`)

CELfiles<-list.files(DATA_DIR)[grep(".CEL$", list.files(DATA_DIR))]
linker<-strsplit(CELfiles, "-") %>% sapply("[[", 1)
linker==directory.ws$`SPECSMAC#`
directory.ws$FileName<-file.path(DATA_DIR, CELfiles)
rownames(directory.ws)<-directory.ws$`SPECSMAC#`
```

```{r build microarray object}
setwd("/fh/fast/furlan_s/user/owalt/ewings/EwingMA/CEL")
rawData <- read.celfiles(CELfiles)
getClass("GeneFeatureSet")

exprs(rawData)[1:4,1:3]
filename <- sampleNames(rawData)
# colData(rawData)$filename <- filename

#filter features by genes
geneset<-oligo::rma(rawData, target = "core")
# BiocManager::install(geneset@annotation)
geneset

geneAcc=as.character(huex10sttranscriptclusterACCNUM[rownames(geneset)])
geneNames=as.character(huex10sttranscriptclusterSYMBOL[rownames(geneset)])

geneset<-annotateEset(geneset,huex10sttranscriptcluster.db) #remove probes that don't have a gene

mat<-Biobase::exprs(geneset)
colnames(mat)<-colnames(mat) %>% strsplit("-")%>% sapply("[[", 1)

out<-read.csv("/fh/fast/furlan_s/user/owalt/ewings/EwingMA/cog_log2_pdata.csv")
rownames(out)<-out$sample %>% gsub(".CEL", "", .)
colnames(mat)
fdat<- geneset@featureData@data
names(fdat)<-c("id", "entrez", "gene_short_name", "description")
obj<-CreateSeuratObject(counts = mat %>%as.sparse(), meta.data = out)
```

```{r}
fus_kmean<-readRDS("~/m3/res/kd_kmean_genes.rds")


kmean_probes<-lapply(fus_kmean, function(x){
  fdat[fdat$gene_short_name %in% x,]$id  
})


obj<-AddModuleScore(obj, features = kmean_probes, name = "k")
obj@meta.data$OS..censor.0..dead.1.<-obj@meta.data$OS..censor.0..dead.1. %>%as.character()
p<-ggplot(obj@meta.data[obj@meta.data$EFS.event != "SMN",], aes(x=EFS.event, y = k2))+geom_boxplot()
p + geom_dotplot(binaxis='y', stackdir='center', dotsize=1,binwidth = 1/30)+stat_compare_means()
```


#Fig 5D
```{r survival curves kmean}
pdf("../figs/k1_survival.pdf", width = 3, height = 3)
strat_curve(obj, gene_list= list(kmean_probes[[1]]),name =  "k1")+scale_color_manual(values =c("cornflowerblue", "#FF4500"))
dev.off()

pdf("../figs/k2_survival.pdf", width = 3, height = 3)
strat_curve(obj, gene_list= list(kmean_probes[[2]]),name =  "k2")+scale_color_manual(values =c("cornflowerblue", "#FF4500"))
dev.off()

pdf("../figs/k3_survival.pdf", width = 3, height =3)
strat_curve(obj, gene_list= list(kmean_probes[[3]]),name =  "k3")+scale_color_manual(values =c("cornflowerblue", "#FF4500"))
dev.off()


pdf("../figs/k3_survival.pdf", width = 3, height =3)
strat_curve(obj, gene_list= list(fosl2_probes[[1]]),name =  "fosl2_targets")+scale_color_manual(values =c("cornflowerblue", "#FF4500"))
dev.off()
```



# Mak PDX object
```{r}
samps<- c("PDX_184", "PDX_352", "PDX_856","PDX_861", "PDX_1058")

seus<-lapply(1:length(samps), function(i){
  files<-file.path(DATA_DIR, "ddata/GSE130024_RAW", samps[i])
  raw<-Read10X(files)
  seu<-CreateSeuratObject(raw)
  seu$dataset<- samps[i]
  seu
})

pdx <- merge(
  x = seus[[1]],
  y = seus[2:length(seus)],
  add.cell.ids = samps
)

pdx@assays[["RNA"]]@data@Dimnames[[1]]<-gsub("GRCh38-", "",pdx@assays[["RNA"]]@data@Dimnames[[1]])
pdx@assays[["RNA"]]@counts@Dimnames[[1]]<-gsub("GRCh38-", "",pdx@assays[["RNA"]]@counts@Dimnames[[1]])
rownames(pdx[,pdx$dataset == "PDX_1058"])


DefaultAssay(pdx) <- "RNA"
pdx[["percent.mt"]] <- PercentageFeatureSet(pdx, pattern = "^MT-")
pdx$log_RNA<-log10(pdx$nCount_RNA)

```

```{r RNA/ATAC QC}
VlnPlot(pdx, features =  c("nCount_RNA", "percent.mt"), pt.size = 0, ncol = 3, log=T, group.by="dataset")
VlnPlot(pdx, features =  c("percent.mt"), pt.size = 0, group.by="dataset", y.max= 50)
VlnPlot(pdx, features =  c("log_RNA"), pt.size = 0, group.by="dataset")

# pdx<-pdx[,pdx$dataset != "PDX_1058"]
```

```{r RNA/ATAC QC}
pdx<- subset(
  x =pdx,
  subset = 
    percent.mt <= 15 &
    log_RNA <= 4.5 &
    log_RNA >= 3 )

pdx <- SCTransform(pdx, vars.to.regress = "percent.mt", verbose = FALSE)
pdx <- RunPCA(pdx, verbose = FALSE)
pdx <- RunUMAP(pdx, dims = 1:30, verbose = FALSE)

pdx <- FindNeighbors(pdx, dims = 1:30, verbose = FALSE)
pdx <- FindClusters(pdx, verbose = FALSE)
DimPlot(pdx, label = TRUE) + NoLegend()
```

# add PDX metadata
```{r}
pdx$localization<-pdx$dataset
pdx$localization[pdx$dataset == "PDX_184"]<-"localized (diagnosis)"
pdx$localization[pdx$dataset == "PDX_352"]<-"localized (diagnosis)"
pdx$localization[pdx$dataset == "PDX_861"]<-"metastatic (diagnosis)"
pdx$localization[pdx$dataset == "PDX_856"]<-"localized (relapse)"
pdx$localization[pdx$dataset == "PDX_1058"]<-"metastatic (relapse)"
```


# PDX Dim Plot, Fig S6C
```{r}
DimPlot(pdx, reduction = "sct_umap", group.by = "dataset")
cds<-seurat_to_monocle3(pdx, seu_rd = "sct_umap")
pdf("~/pdx/figs/dataset_umap.pdf", width = 5, height = 4)
plot_cells(cds, color_cells_by = "dataset", label_cell_groups = F)+scale_color_manual(values = pal_discrete(values = levels(factor(cds$dataset)), set = "calm", reverse = T))&NoAxes()
dev.off()
```

# kmean analysis, Fig S6D
```{r fusion regulated kmean genes}
fus_kmean<-readRDS("~/m3/res/kd_kmean_genes.rds")
pdx<-AddModuleScore(pdx, features = fus_kmean, name = "k")

mat<-matrix(nrow = 3, ncol = length(colnames( pdx)))
rownames(mat)<-c("k1", "k2", "k3")
colnames(mat)<-colnames( pdx)
mat[1,]<- pdx@meta.data$k1
mat[2,]<- pdx@meta.data$k2
mat[3,]<- pdx@meta.data$k3

mat<- t(scale(t(mat)))
id<-factor( pdx$dataset)

col_fun = circlize::colorRamp2(quantile(mat, c(0.01, 0.25, 0.5, 0.75, 0.99)), paletteContinuous(n=5))
pdf("~/pdx/figs/fus_kmean_heatmap.pdf", width = 11, height = 1.5)
Heatmap(mat, name = "Kmean_Expression",cluster_rows = T, cluster_columns = T, show_row_dend = F,row_names_rot = 0, column_names_rot = -40, show_column_names = F, column_split = id, show_column_dend = F, use_raster = F, col = col_fun, clustering_distance_columns = "pearson", clustering_distance_rows = "pearson",top_annotation = HeatmapAnnotation(foo = anno_block(gp = gpar(fill = pdx_cols, col = 0))))
dev.off()
```



