---
title: Figure 1, Global Ewing Sarcoma signatures
author: OW
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
rm(list=ls())
suppressPackageStartupMessages({
  library(monocle3)
  library(reticulate)
  library(openxlsx)  
  library(dplyr)
  library(Matrix)
  library(ggplot2)
  library(xfun)
  library(pals)
  library(rhdf5)
  library(RColorBrewer)
  library(Signac)
  library(Seurat)
  library(ggplot2)
  library(future)
  library(GenomicRanges)
  library(ComplexHeatmap)
  library(EnsDb.Hsapiens.v86)
  library(viridis)
  library(BSgenome.Hsapiens.UCSC.hg38)
  library(JASPAR2020)
  library(TFBSTools)
  library(ArchR)
  library(parallel)
  library(scCustomize)
  library(SeuratObject)
})
library(sp)
library(stringi)
BiocManager::install("stringi")

Sys.setenv(MODULEPATH="/app/modules/all", MODULEPATH_ROOT="/app/modules/all", MODULESHOME="/app/lmod/lmod")
source("/app/lmod/lmod/init/R", echo=FALSE, max=Inf)
module("load", "MACS2/2.2.6-foss-2019b-Python-3.7.4")

Sys.setenv(R_LIBS_USER = "/home/owaltner/R/x86_64-pc-linux-gnu-library/4.2")
Sys.getenv()
#version 11 cuda
dyn.load('/app/software/ArrayFire/3.8.1/lib64/libaf.so.3')

knitr::opts_chunk$set(dev="CairoPNG")

library(RcppArrayFire)
library(viewmastR)
```

#load objects
```{r}
set.seed(1234)

setwd("/fh/fast/furlan_s/user/owalt/ewings/multiome/all_parents")
m1<- loadArchRProject("/fh/fast/furlan_s/user/owalt/ewings/multiome/all_parents")
m1_seu<-readRDS("/fh/fast/furlan_s/user/owalt/ewings/multiome/all_parents/all_parents_seurat.rds")

setwd("/fh/fast/furlan_s/user/owaltner/m2") 
m2<-loadArchRProject()
m2_seu<-readRDS("m2.rds")
```

# colors
```{r}
all_parents_cols<-c("A4573"="#D51F26","A673"="#272E6A","CHLA9"="#208A42","CHLA10"="#89288F","PDX305"="#F47D2B", "RDES"="#FEE500","SKNMC"="#8A9FD1","TC32"="#3BBCA8","TC71"="#E6C2DC",
               "hMSC_parent_240"="#90D5E4",  "hMSC_parent"="#89C75F","hMSC.h7"="#F37B7D","RD_parent"="#78385c","U2OS_parent"="#C06CAB")

cell_line_cols<-c( "A673" = "#272E6A", "CHLA9"= "#208A42", "CHLA10" = "#89288F", "TC32" = "#00bca8", "PDX305"= "#F47D2B", "SKNMC"= "#8A9FD1", "A4573" = "#D51F26", "TC71" = "#e9c2db","RDES" = "#FEE500")
subgroup_cols<-c("group.1"= "#9d8cbf","group.2" = "#12733c","group.3"= "#edd30c")

cluster_cols<-c( "C1" = "#2d3aad",
                 "C2" = "#01031f",
                 "C3" = "#D51F26",
                 "C4" = "#d64b50",
                 "C5" = "#78181b",
                 "C6" = "#8ed1f5",
                 "C7" = "#739cff",
                 "C8" = "#5e5cab",
                 "C9" = "#abbeed",
                 "C10" = "#8A9FD1",
                 "C11" = "#e9c2db",
                 "C12" = "#d17bb2",
                 "C13" = "#FEE500",
                 "C14" = "#bfb341",
                 "C15" = "#ded8a4",
                 "C16" = "#00bca8",
                 "C17" = "#10665d",
                 "C18" = "#77a39f",
                 "C19" = "#8a3c07",
                 "C20" = "#eb9d67",
                 "C21" = "#F47D2B",
                 "C22" = "#89288F",
                 "C23" = "#b452ba",
                 "C24" = "#208a42",
                 "C25" = "#06541f")

rna_cols<-paletteContinuous(n=8)[c(1:3, 6:8)]

pal<- rev(mako(n=9))
pal<-pal[1:8]
pal<-c("#ffffff", pal)
mtf_cols<-pal

atac_cols <- paletteContinuous(set = "blueYellow",n = 10)
```

# Dimplot, Fig 1A
```{r reduce dimensions }
cds<-seurat_to_monocle3(m1_seu, seu_rd = "mo_UMAP", assay = "MotifMatrix")
plot_cells(cds, color_cells_by = "cell_line", cell_size = 0.65, label_cell_groups = F)+scale_color_manual(values = c("A4573"="#D51F26","A673"="#272E6A","CHLA9"="#208A42","CHLA10"="#89288F","PDX305"="#F47D2B", "RDES"="#FEE500","SKNMC"="#8A9FD1","TC32"="#3BBCA8","TC71"="#E6C2DC",
               "hMSC_parent_240"="#90D5E4",  "hMSC_parent"="#89C75F","hMSC.h7"="#F37B7D","RD_parent"="#78385c","U2OS_parent"="#C06CAB"))&NoAxes()
```

# EWSFLI1 motif, Fig 1B
```{r EWSFLI1 motif plot}
m1$cell_line<-as.character(m1$cell_line)     
p <- plotGroups(ArchRProj = m1, 
  groupBy = "cell_line", 
  colorBy = "MotifMatrix", 
  name = "z:EWSR1.FLI1_22",
  imputeWeights = getImputeWeights(m1),
  threads =1,
  groupOrder = rev(c("SKNMC","A4573","CHLA10", "RDES","PDX305","TC32","CHLA9","A673","TC71","U2OS_parent", "hMSC.h7","hMSC_parent","RD_parent","hMSC_parent_240")))


p2<-p+scale_fill_manual(values = c("A4573"="#D51F26","A673"="#272E6A","CHLA9"="#208A42","CHLA10"="#89288F","PDX305"="#F47D2B", "RDES"="#FEE500","SKNMC"="#8A9FD1","TC32"="#3BBCA8","TC71"="#E6C2DC",
               "hMSC_parent_240"="#90D5E4",  "hMSC_parent"="#89C75F","hMSC.h7"="#F37B7D","RD_parent"="#78385c","U2OS_parent"="#C06CAB"))


plotPDF(p2, name = "EWS_chromvar_ridgeplot", width =7, height = 7, ArchRProj = m1, addDOC = FALSE)

```

#GGAA microsatelitie EWS:FLI binding sites chromvar Fig S1D
```{r}
msat_regs<-read.table("/fh/fast/furlan_s/user/owalt/ewings/multiome/9cl/res/ggaa.bed", header = F)
non_msat_regs<-read.table("/fh/fast/furlan_s/user/owalt/ewings/multiome/9cl/res/non_ggaa.bed")

names(msat_regs)<-c("chr", "start", "end", "score", "strand")
names(non_msat_regs)<-c("chr", "start", "end", "score", "strand")

msat_gr<-makeGRangesFromDataFrame(msat_regs)
non_msat_gr<-makeGRangesFromDataFrame(non_msat_regs)
  
m1<-addPeakAnnotations(ArchRProj = m1, regions = list("msat"=msat_gr,"non_msat"=non_msat_gr), name = "msat_EWS:FLI_binding_site", force =T)
m1 <- addDeviationsMatrix(
  ArchRProj = m1, 
  peakAnnotation = "msat_EWS:FLI_binding_site",
  force = TRUE
)
getAvailableMatrices(m1)

markerChIP <- getFeatures(m1, useMatrix = "msat_EWS:FLI_binding_siteMatrix")
markerChIP <- sort(grep("z:", markerChIP, value = TRUE))
markerChIP

p <- plotEmbedding(
    ArchRProj = m1, 
    colorBy = "msat_EWS:FLI_binding_siteMatrix", 
    name = markerChIP, 
    embedding = "mo_UMAP"
)

msat_bind_mat<-getMatrixFromProject(m1, useMatrix = "msat_EWS:FLI_binding_siteMatrix")

msat_bind_mat<-assays(msat_bind_mat)$deviations

t<-t(msat_bind_mat)

t

m1_seu@meta.data$msat_EWS_FLI_binding<-t[,1]
m1_seu@meta.data$non_msat_EWS_FLI_binding<-t[,2]

dim(msat_bind_mat)
dim(m1_seu)

p<-FeaturePlot_scCustom(m1_seu, features = c("msat_EWS_FLI_binding", "non_msat_EWS_FLI_binding"), order = T, pt.size = 0.1, max.cutoff = "q99")&NoAxes()
ggsave(p, filename="/fh/fast/furlan_s/user/owalt/ewings/multiome/all_parents/Plots/ews_fli_bind_msat_umap.pdf", width =10, height =5)
```

# GGAA microsatelitie EWS:FLI binding sites chromvar Fig 1C, Fig S1 D
```{r}
msat_regs<-read.table("/fh/fast/furlan_s/user/owalt/ewings/multiome/9cl/res/ggaa.bed", header = F)
non_msat_regs<-read.table("/fh/fast/furlan_s/user/owalt/ewings/multiome/9cl/res/non_ggaa.bed")

names(msat_regs)<-c("chr", "start", "end", "score", "strand")
names(non_msat_regs)<-c("chr", "start", "end", "score", "strand")

msat_gr<-makeGRangesFromDataFrame(msat_regs)
non_msat_gr<-makeGRangesFromDataFrame(non_msat_regs)
  
m1<-addPeakAnnotations(ArchRProj = m1, regions = list("msat"=msat_gr,"non_msat"=non_msat_gr), name = "msat_EWS:FLI_binding_site", force =T)
m1 <- addDeviationsMatrix(
  ArchRProj = m1, 
  peakAnnotation = "msat_EWS:FLI_binding_site",
  force = TRUE
)
getAvailableMatrices(m1)

markerChIP <- getFeatures(m1, useMatrix = "msat_EWS:FLI_binding_siteMatrix")
markerChIP <- sort(grep("z:", markerChIP, value = TRUE))
markerChIP

p <- plotEmbedding(
    ArchRProj = m1, 
    colorBy = "msat_EWS:FLI_binding_siteMatrix", 
    name = markerChIP, 
    embedding = "mo_UMAP"
)

msat_bind_mat<-getMatrixFromProject(m1, useMatrix = "msat_EWS:FLI_binding_siteMatrix")

msat_bind_mat<-assays(msat_bind_mat)$deviations

t<-t(msat_bind_mat)

t

m1_seu@meta.data$msat_EWS_FLI_binding<-t[,1]
m1_seu@meta.data$non_msat_EWS_FLI_binding<-t[,2]

dim(msat_bind_mat)
dim(m1_seu)

FeaturePlot_scCustom(m1_seu, features = c("msat_EWS_FLI_binding", "non_msat_EWS_FLI_binding"), order = T, colors_use = mtf_cols[c(-1, -9)], pt.size = 0.7, max.cutoff = "q99")&NoAxes()

VlnPlot(m1_seu, group.by ="cell_line", features = c("msat_EWS_FLI_binding", "non_msat_EWS_FLI_binding"), pt.size = 0)+scale_fill_manual(values = all_parents_cols)
```

# Fig 1D
```{r}
combined_df<-read.csv("/fh/fast/furlan_s/user/owalt/ewings/fusion/pb_fusion_results.csv")
   
#load ewing + non ewing object
m1<-readRDS("all_parents_seurat.rds")
m1<-m1_seu


combined_df$cb
colnames(m1)

colnames(m1)<-gsub("#", "_", colnames(m1))

table(colnames(m1) %in% combined_df$cb)
length(colnames(m1))

fus_df<-combined_df[grepl("EWSR1", combined_df$fusion) & grepl("FLI1", combined_df$fusion),]
mat<-table(colnames(m1) %in% fus_df$cb, m1$cell_line) %>% as.matrix()
mat


df<-table(colnames(m1) %in% fus_df$cb, m1$cell_line) %>% as.data.frame()

meta_fus<-fus_df[!duplicated(fus_df$cb),]

m1$fus<-"NA"
m1$fus[which(colnames(m1) %in% meta_fus$cb)]<-"EWSR1_FLI1"

pb_res<-table(m1$fus, m1$cell_line)


m1$pos_fus<-ifelse(m1$fus >0, "pos", "neg")
df<-table(m1$pos_fus, m1$cell_line) %>% as.data.frame() %>% group_by(Var2)
df<-df[df$Var1 == "pos",] %>% arrange(Freq)
df$Var2<-factor(df$Var2, levels =  df$Var2)
ggplot(df[df$Var1 == "pos",], aes(x=Var2, y = Freq , fill =Var2))+geom_bar(stat = "identity", color = "black")+scale_fill_manual(values = all_parents_cols)
```

#Fig S1E
```{r}
DimPlot(m1, group.by = "fus")+scale_color_manual(values = c("red", "gray90"))

cds<-viewmastR::seurat_to_monocle3(m1, seu_rd = "mo_UMAP")

pdf("/fh/fast/furlan_s/user/owalt/ewings/fusion/pbfusion_umap.pdf", width =6, height=5)
plot_cells(cds[,order(cds$fus, decreasing = T)], color_cells_by = "fus", cell_size = 0.75, label_cell_groups = F)+scale_color_manual(values = c("red", "gray90"))+theme( panel.background = element_rect(fill='transparent'), #transparent panel bg
    plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
    panel.grid.major = element_blank(), #remove major gridlines
    panel.grid.minor = element_blank(), #remove minor gridlines
    legend.background = element_rect(fill='transparent'), #transparent legend bg
    text = element_text(color = "white") )&NoAxes()

dev.off()
```


# Top Markers Genes Heatmap, 1E
```{r}
markersGenes <- getMarkerFeatures(
    ArchRProj = m1, 
    useMatrix = "GeneExpressionMatrix", 
    groupBy = "cell_line",
    bias = c( "Gex_nUMI"),
    testMethod = "wilcoxon"
)

heatmapGenes <- markerHeatmap(
  seMarker = markersGenes, 
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25", 
  transpose = TRUE,nLabel = 0
  
)

ComplexHeatmap::draw(heatmapGenes, heatmap_legend_side = "bot", annotation_legend_side = "bot")
```


# Top Peaks Heatmap, Fig 1F
```{r}
markersPeaks <- getMarkerFeatures(
    ArchRProj = m1, 
    useMatrix = "PeakMatrix", 
    groupBy = "cell_line",
    testMethod = "wilcoxon"
)

heatmapPeaks <- markerHeatmap(
  seMarker = markersPeaks, 
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25", 
  transpose = TRUE,nLabel = 0
  
)

ComplexHeatmap::draw(heatmapGenes, heatmap_legend_side = "bot", annotation_legend_side = "bot")
```


#  qc stats Fig S1 B
```{r}
all_parents_cols<-c("A4573"="#D51F26","A673"="#272E6A","CHLA9"="#208A42","CHLA10"="#89288F","PDX305"="#F47D2B", "RDES"="#FEE500","SKNMC"="#8A9FD1","TC32"="#3BBCA8","TC71"="#E6C2DC",
               "hMSC_parent_240"="#90D5E4",  "hMSC_parent"="#89C75F","hMSC.h7"="#F37B7D","RD_parent"="#78385c","U2OS_parent"="#C06CAB")


pdf("/fh/fast/furlan_s/user/owalt/ewings/multiome/qc/m1_qc.pdf", width = 20, height = 4.5)
VlnPlot(m1_seu, features = c("Gex_nGenes", "Gex_nUMI", "TSSEnrichment", "ReadsInBlacklist", "nFrags"), group.by = "cell_line", cols = all_parents_cols, pt.size= 0, ncol = 5)
dev.off()

```


# IC_EWS, Supp Fig C
```{r Gene expression of IC_EWS}
aynaud<- read.delim(file = "/fh/fast/furlan_s/user/owalt/ewings/gene_sets/ICEWS_genes.txt")
DefaultAssay(m1_seu)<-"GeneExpressionMatrix"
m1_seu<-NormalizeData(m1_seu)

m1_seu<-AddModuleScore(m1_seu, features = list(aynaud$Gene), name = "Aynaud")

Idents(m1_seu)<-"cell_line"
m1_seu$cell_line<-factor(m1_seu$cell_line, levels = c("hMSC_parent_240",  "hMSC_parent", "hMSC.h7", "RD_parent", "U2OS_parent","TC71","A673", "CHLA9", "TC32", "PDX305","RDES","CHLA10","A4573",   "SKNMC"))

pdf("aynaud_vln2.pdf", width = 12, height = 5)
VlnPlot(m1_seu, features = "Aynaud1",  pt.size = 0,  cols = all_parents_cols,group.by = "cell_line")
dev.off()
```

```{r Gene Activity of IC_EWS}
DefaultAssay(m1_seu)<-"GeneScoreMatrix"

m1_seu<-AddModuleScore(m1_seu, features = list(aynaud$Gene), name = "Aynaud_GS")

pdf("aynaud_gene_acc.pdf", width = 12, height = 5)
VlnPlot(m1_seu, features = "Aynaud_GS1",  pt.size = 0,  cols = all_parents_cols,group.by = "cell_line")&NoLegend()
dev.off()
```


